File: EduLink.Domain/Entities/States/ConfirmadaState.cs
using EduLink.Domain.Enums;
using System;

namespace EduLink.Domain.Entities.States;

public class ConfirmadaState : ReservaState
{
    public override EstadoReserva Nombre => EstadoReserva.Confirmada;

    public override void Cancelar(Reserva reserva, DateTime ahora)
    {
        if (!reserva.PoliticaCancelacion.PuedeCancelar(reserva.Slot.Inicio, ahora))
            throw new InvalidOperationException("Cancelación fuera de plazo. Aplica cargo.");

        reserva.EstadoInterno = new CanceladaState();
        reserva.Slot.Cancelar();
    }

    public override void Completar(Reserva reserva)
    {
        reserva.EstadoInterno = new CompletadaState();
    }
}

File: app/domain/entities/states/confirmada_state.rb
require_relative './reserva_state'
require_relative '../../enums/estado_reserva' # Corrected path
require_relative './cancelada_state'
require_relative './completada_state'

module Entities
  module States
    class ConfirmadaState < ReservaState
      def nombre
        Enums::EstadoReserva::CONFIRMADA
      end

      def cancelar(reserva, ahora)
        unless reserva.politica_cancelacion.puede_cancelar(reserva.slot.inicio, ahora)
          raise StandardError, 'Cancelación fuera de plazo. Aplica cargo.'
        end

        reserva.estado_interno = CanceladaState.new
        reserva.slot.cancelar
      end

      def completar(reserva)
        reserva.estado_interno = CompletadaState.new
      end
    end
  end
end
